src/wifipaf/WiFiPAFLayer.cpp

```diff
@@ -240,27 +240,49 @@ CHIP_ERROR WiFiPAFLayer::Init(chip::System::Layer * systemLayer)
     return CHIP_NO_ERROR;
 }
 
-void WiFiPAFLayer::Shutdown(OnCancelDeviceHandle OnCancelDevice)
+void WiFiPAFLayer::Shutdown()
 {
-    ChipLogProgress(WiFiPAF, "WiFiPAF: Closing all WiFiPAF sessions to shutdown");
+    CloseAllConnections();
+    mSystemLayer = nullptr;
+    memset(&sWiFiPAFEndPointPool, 0, sizeof(sWiFiPAFEndPointPool));
+    return;
+}
+
+void WiFiPAFLayer::CloseAllConnections()
+{
+    ChipLogProgress(WiFiPAF, "WiFiPAF: Closing all WiFiPAF sessions");
     uint8_t i;
-    WiFiPAFSession * pPafSession;
 
     for (i = 0; i < WIFIPAF_LAYER_NUM_PAF_ENDPOINTS; i++)
     {
-        pPafSession = &mPafInfoVect[i];
-        if (pPafSession->id == UINT32_MAX)
+        WiFiPAFSession * pPafSession = &mPafInfoVect[i];
+        if (pPafSession->id == kUndefinedWiFiPafSessionId)
         {
             // Unused session
             continue;
         }
-        ChipLogProgress(WiFiPAF, "WiFiPAF: Canceling id: %u", pPafSession->id);
-        OnCancelDevice(pPafSession->id, pPafSession->role);
-        WiFiPAFEndPoint * endPoint = sWiFiPAFEndPointPool.Find(reinterpret_cast<WIFIPAF_CONNECTION_OBJECT>(pPafSession));
-        if (endPoint != nullptr)
-        {
-            endPoint->DoClose(kWiFiPAFCloseFlag_AbortTransmission, WIFIPAF_ERROR_APP_CLOSED_CONNECTION);
-        }
+        CloseConnection(PafInfoAccess::kAccSessionId, *pPafSession);
+    }
+}
+
+void WiFiPAFLayer::CloseConnection(PafInfoAccess accType, const WiFiPAFSession & PafSession)
+{
+    WiFiPAFSession * pPafSession = GetPAFInfo(accType, PafSession);
```

Comment:
So when used from CloseAllConnections this leads to O(N^2) behavior.

Maybe that's OK given the small numbers, but as a followup it does seem like it would be clearer if we had:

1.  A public method that takes some session identifier (maybe not a full WiFiPAFSession as discussed above).
2. A private method that takes a WiFiPAFSession.  The public method would GetPAFInfo and call this.
3. The private method would be called directly from CloseAllConnections, since we know we have a "canonical" WiFiPAFSession there.